testing/extracttest/extracttest.go

```diff
@@ -0,0 +1,132 @@
+// Package extracttest provides structures to help create tabular tests for extractors.
+package extracttest
+
+import (
+	ordercmp "cmp"
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"slices"
+	"strings"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	"github.com/google/osv-scalibr/extractor"
+	"github.com/google/osv-scalibr/extractor/filesystem"
+	scalibrfs "github.com/google/osv-scalibr/fs"
+	"github.com/google/osv-scalibr/testing/fakefs"
+)
+
+// ScanInputMockConfig is used to quickly configure building a mock ScanInput
+type ScanInputMockConfig struct {
+	// Path of the file ScanInput will read, relative to the ScanRoot
+	Path string
+	// FakeScanRoot allows you to set a custom scanRoot, can be relative or absolute,
+	// and will be translated to an absolute path
+	FakeScanRoot string
+	FakeFileInfo *fakefs.FakeFileInfo
+}
+
+// TestTableEntry is a entry to pass to ExtractionTester
+type TestTableEntry struct {
+	Name          string
+	InputConfig   ScanInputMockConfig
+	WantInventory []*extractor.Inventory
+	WantErr       error
+}
+
+// ContainsErrStr is an error that matches other errors that contains
+// `str` in their error string.
+type ContainsErrStr struct {
+	Str string
+}
+
+// Error returns the error string
+func (e ContainsErrStr) Error() string { return fmt.Sprintf("error contains: '%s'", e.Str) }
+
+// Is checks whether the input error contains the string in ContainsErrStr
+func (e ContainsErrStr) Is(err error) bool {
+	return strings.Contains(err.Error(), e.Str)
+}
+
+// ExtractionTester tests common properties of a extractor, and returns the raw values from running extract
+func ExtractionTester(t *testing.T, extr filesystem.Extractor, tt TestTableEntry) ([]*extractor.Inventory, error) {
+	t.Helper()
+
+	wrapper := generateScanInputMock(t, tt.InputConfig)
+	got, err := extr.Extract(context.Background(), &wrapper.ScanInput)
+	wrapper.close()
+
+	if !cmp.Equal(err, tt.WantErr, cmpopts.EquateErrors()) {
+		t.Errorf(
+			"%s.Extract(%s) error diff:\n%s",
+			extr.Name(),
+			tt.InputConfig.Path,
+			cmp.Diff(err, tt.WantErr, cmpopts.EquateErrors()),
+		)
+		return got, err
+	}
+
+	// Sort the results before comparing to ensure same order.
+	sortFunc := func(a, b *extractor.Inventory) int {
```

Comment:
One thing Victor pointed out is that one can just specify a SortOption in cmp.Diff so we might not need this (and we can reuse the existing comparison function from scalibr/scalibr.go for this.)

If we removed the sorting I'm not sure if there'd be that much code left - I think the comparison would look something like the one in this PR: https://github.com/google/osv-scalibr/pull/175/files#diff-cef30953297dcf14e72625ac3e41a4e3eaf75106391a0e4c06ae35c04ba49ec0R167

Given that there'd only be 9 lines of code left maybe we should follow https://google.github.io/styleguide/go/best-practices in that case and inline those in the tests instead of using a separate helper lib.

Reply:
Refactored now according to offline discussion, removed the helper lib and moved the cmp and t.Errorf calls directly into the tests.